File: recipes/hostui/gudule-greeting-hostui-compose/src/main/kotlin/greeting/gokotlin/grpc/DaemonProcess.kt
package greeting.gokotlin.grpc

import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import kotlinx.coroutines.runBlocking
import kotlin.io.path.absolutePathString
import kotlin.io.path.exists
import org.organicprogramming.holons.Connect

class DaemonProcess {
    companion object {
        private const val holonSlug = "greeting-daemon-greeting-gokotlin"
        private const val holonUUID = "1a409a1e-69e3-4846-9f9b-47b0a6f98f84"
        private const val familyName = "Greeting-Gokotlin"
    }

    private var client: GreetingClient? = null
    private var stageRoot: Path? = null

    fun client(): GreetingClient {
        if (client == null) {
            start()
        }
        return requireNotNull(client)
    }

    fun start() {
        if (client != null) {
            return
        }

        val daemonPath = resolveDaemonPath()
        val root = stageHolonRoot(daemonPath)
        stageRoot = root

        val previousUserDir = System.getProperty("user.dir")
        try {
            System.setProperty("user.dir", root.toString())
            val channel = runBlocking { Connect.connect(holonSlug) }
            client = GreetingClient(channel)
        } catch (t: Throwable) {
            stageRoot = null
            deleteStageRoot(root)
            throw t
        } finally {
            if (previousUserDir == null) {
                System.clearProperty("user.dir")
            } else {
                System.setProperty("user.dir", previousUserDir)
            }
        }
    }

    fun stop() {
        val root = stageRoot
        stageRoot = null
        val activeClient = client
        client = null

        try {
            activeClient?.close()
        } finally {
            if (root != null) {
                deleteStageRoot(root)
            }
        }
    }

    private fun resolveDaemonPath(): String {
        val devBinary = Path.of("../greeting-daemon/gudule-daemon-greeting-gokotlin")
        if (devBinary.exists()) {
            return devBinary.normalize().absolutePathString()
        }

        val resource = javaClass.classLoader.getResourceAsStream("embedded/gudule-daemon-greeting-gokotlin")
            ?: error("Embedded daemon resource is missing")

        val tempFile = Files.createTempFile("gudule-daemon-gokotlin", "")
        resource.use { input ->
            Files.copy(input, tempFile, java.nio.file.StandardCopyOption.REPLACE_EXISTING)
        }
        File(tempFile.toUri()).setExecutable(true)
        tempFile.toFile().deleteOnExit()
        return tempFile.absolutePathString()
    }

    private fun stageHolonRoot(binaryPath: String): Path {
        val root = Files.createTempDirectory("greeting-gokotlin-holon-")
        val holonDir = root.resolve("holons").resolve(holonSlug)
        Files.createDirectories(holonDir)
        Files.writeString(holonDir.resolve("holon.yaml"), manifestFor(binaryPath))
        return root
    }

    private fun manifestFor(binaryPath: String): String = """
schema: holon/v0
uuid: "$holonUUID"
given_name: greeting-daemon
family_name: "$familyName"
motto: Greets users in 56 languages — a Gokotlin recipe example.
composer: B. ALTER
clade: deterministic/pure
status: draft
born: "2026-02-20"
generated_by: manual
kind: native
build:
  runner: go-module
artifacts:
  binary: "${yamlEscape(binaryPath)}"
""".trimIndent() + "\n"

    private fun yamlEscape(value: String): String =
        value.replace("\\", "\\\\").replace("\"", "\\\"")

    private fun deleteStageRoot(root: Path) {
        runCatching {
            val directory = root.toFile()
            if (directory.exists()) {
                directory.deleteRecursively()
            }
        }
    }
}

File: recipes/hostui/gudule-greeting-hostui-dotnet/Services/DaemonProcess.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using Grpc.Net.Client;
using Holons;

namespace Greeting.Godotnet.Services;

public sealed class DaemonProcess : IAsyncDisposable
{
    private readonly string _binaryName;
    private GrpcChannel? _channel;
    private GreetingClient? _client;
    private string? _stageRoot;

    public DaemonProcess(string binaryName)
    {
        _binaryName = binaryName;
    }

    public string HolonSlug => "greeting-daemon-greeting-godotnet";

    public GreetingClient Client =>
        _client ?? throw new InvalidOperationException("Daemon is not connected.");

    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        if (_client is not null)
        {
            return;
        }

        AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
        var binaryPath = await ResolveBinaryPathAsync(cancellationToken);
        var stageRoot = await StageHolonRootAsync(binaryPath, cancellationToken);
        _stageRoot = stageRoot;

        var previousDirectory = Directory.GetCurrentDirectory();
        try
        {
            cancellationToken.ThrowIfCancellationRequested();
            Directory.SetCurrentDirectory(stageRoot);
            _channel = Connect.ConnectTarget(HolonSlug);
            _client = new GreetingClient(_channel);
        }
        catch
        {
            if (_channel is not null)
            {
                Connect.Disconnect(_channel);
                _channel = null;
            }

            _client = null;
            _stageRoot = null;
            await DeleteStageRootAsync(stageRoot);
            throw;
        }
        finally
        {
            Directory.SetCurrentDirectory(previousDirectory);
        }
    }

    public async ValueTask DisposeAsync()
    {
        _client = null;
        var root = _stageRoot;
        _stageRoot = null;

        try
        {
            if (_channel is not null)
            {
                Connect.Disconnect(_channel);
                _channel = null;
            }
        }
        finally
        {
            if (root is not null)
            {
                await DeleteStageRootAsync(root);
            }
        }
    }

    private async Task<string> ResolveBinaryPathAsync(CancellationToken cancellationToken)
    {
        var packagedPath = await TryResolvePackagedBinaryAsync(cancellationToken);
        if (packagedPath is not null)
        {
            return packagedPath;
        }

        var devPath = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "greeting-daemon", _binaryName));
        if (File.Exists(devPath))
        {
            return devPath;
        }

        throw new FileNotFoundException($"Unable to locate {_binaryName} in the app bundle or source tree.");
    }

    private async Task<string?> TryResolvePackagedBinaryAsync(CancellationToken cancellationToken)
    {
        var cachePath = Path.Combine(FileSystem.Current.CacheDirectory, _binaryName);
        if (File.Exists(cachePath))
        {
            return cachePath;
        }

        try
        {
            await using var packageStream = await FileSystem.Current.OpenAppPackageFileAsync(_binaryName);
            await using var output = File.Create(cachePath);
            await packageStream.CopyToAsync(output, cancellationToken);
            await output.FlushAsync(cancellationToken);
            await MakeExecutableAsync(cachePath, cancellationToken);
            return cachePath;
        }
        catch (FileNotFoundException)
        {
            return null;
        }
    }

    private static async Task MakeExecutableAsync(string path, CancellationToken cancellationToken)
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return;
        }

        var chmod = new ProcessStartInfo
        {
            FileName = "/bin/chmod",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        chmod.ArgumentList.Add("+x");
        chmod.ArgumentList.Add(path);

        using var process = Process.Start(chmod) ?? throw new InvalidOperationException("Unable to invoke chmod.");
        await process.WaitForExitAsync(cancellationToken);
    }

    private async Task<string> StageHolonRootAsync(string binaryPath, CancellationToken cancellationToken)
    {
        var root = Path.Combine(
            FileSystem.Current.CacheDirectory,
            "holons-runtime",
            HolonSlug);
        var holonDir = Path.Combine(root, "holons", HolonSlug);
        Directory.CreateDirectory(holonDir);
        await File.WriteAllTextAsync(
            Path.Combine(holonDir, "holon.yaml"),
            BuildManifest(binaryPath),
            cancellationToken);
        return root;
    }

    private string BuildManifest(string binaryPath)
    {
        var escapedPath = binaryPath.Replace("\\", "\\\\", StringComparison.Ordinal)
            .Replace("\"", "\\\"", StringComparison.Ordinal);
        return $"""
schema: holon/v0
uuid: "1a409a1e-69e3-4846-9f9b-47b0a6f98f84"
given_name: greeting-daemon
family_name: "Greeting-Godotnet"
motto: Greets users in 56 languages — a Go + .NET recipe example.
composer: Codex
clade: deterministic/pure
status: draft
born: "2026-03-06"
generated_by: manual
kind: native
build:
  runner: go-module
artifacts:
  binary: "{escapedPath}"
""" + Environment.NewLine;
    }

    private static Task DeleteStageRootAsync(string root)
    {
        if (!Directory.Exists(root))
        {
            return Task.CompletedTask;
        }

        return Task.Run(() => Directory.Delete(root, recursive: true));
    }
}




